File: Pekkam-iOS/Pekkam/ImageOverlayRenderer.swift
import UIKit

struct ImageOverlayRenderer {
    static func addWatermark(to image: UIImage) -> UIImage {
        let renderer = UIGraphicsImageRenderer(size: image.size)
        
        let result = renderer.image { ctx in
            image.draw(at: .zero)
            
            let text = "PEKKAM PRØVEVERSJON"
            let attributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 40, weight: .bold),
                .foregroundColor: UIColor.white.withAlphaComponent(0.3),
            ]
            
            let textSize = text.size(withAttributes: attributes)
            let angle: CGFloat = -45 * .pi / 180
            let spacing: CGFloat = 150
            
            ctx.cgContext.setFillColor(UIColor.clear.cgColor)
            ctx.cgContext.saveGState()
            
            var x = -image.size.width
            while x < image.size.width * 2 {
                var y = -image.size.height
                while y < image.size.height * 2 {
                    ctx.cgContext.saveGState()
                    
                    let centerX = x + textSize.width / 2
                    let centerY = y + textSize.height / 2
                    
                    ctx.cgContext.translateBy(x: centerX, y: centerY)
                    ctx.cgContext.rotate(by: angle)
                    ctx.cgContext.translateBy(x: -centerX, y: -centerY)
                    
                    (text as NSString).draw(at: CGPoint(x: x, y: y), withAttributes: attributes)
                    
                    ctx.cgContext.restoreGState()
                    
                    y += spacing
                }
                x += spacing
            }
            
            ctx.cgContext.restoreGState()
        }
        
        return result
    }
    
    static func addCompassOverlay(to image: UIImage, heading: Double, direction: String) -> UIImage {
        let renderer = UIGraphicsImageRenderer(size: image.size)
        let compassSize: CGFloat = 150
        let padding: CGFloat = 20
        
        let result = renderer.image { ctx in
            image.draw(at: .zero)
            
            let compassRect = CGRect(
                x: image.size.width - compassSize - padding,
                y: image.size.height - compassSize - padding,
                width: compassSize,
                height: compassSize
            )
            
            // Draw background circle
            UIColor.black.withAlphaComponent(0.5).setFill()
            UIBezierPath(ovalIn: compassRect).fill()
            
            // Draw border
            UIColor.white.setStroke()
            let borderPath = UIBezierPath(ovalIn: compassRect)
            borderPath.lineWidth = 2
            borderPath.stroke()
            
            let centerX = compassRect.midX
            let centerY = compassRect.midY
            let radius = compassSize / 2 - 10
            
            // Draw cardinal directions
            let directions = ["N", "E", "S", "W"]
            let angles: [CGFloat] = [0, CGFloat.pi / 2, CGFloat.pi, CGFloat.pi * 1.5]
            
            for (idx, dir) in directions.enumerated() {
                let angle = angles[idx]
                let x = centerX + sin(angle) * radius
                let y = centerY - cos(angle) * radius
                
                let textAttributes: [NSAttributedString.Key: Any] = [
                    .font: UIFont.boldSystemFont(ofSize: 14),
                    .foregroundColor: UIColor.white,
                ]
                
                let textSize = dir.size(withAttributes: textAttributes)
                (dir as NSString).draw(
                    at: CGPoint(x: x - textSize.width / 2, y: y - textSize.height / 2),
                    withAttributes: textAttributes
                )
            }
            
            // Draw needle pointing to heading
            let needleAngle = (heading * .pi / 180)
            let needleLength = radius * 0.8
            let needleEndX = centerX + sin(needleAngle) * needleLength
            let needleEndY = centerY - cos(needleAngle) * needleLength
            
            UIColor.red.setStroke()
            let needlePath = UIBezierPath()
            needlePath.move(to: CGPoint(x: centerX, y: centerY))
            needlePath.addLine(to: CGPoint(x: needleEndX, y: needleEndY))
            needlePath.lineWidth = 3
            needlePath.stroke()
            
            // Draw heading degree text
            let headingText = String(format: "%.0f°", heading)
            let headingAttributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.boldSystemFont(ofSize: 12),
                .foregroundColor: UIColor.white,
            ]
            let headingTextSize = headingText.size(withAttributes: headingAttributes)
            (headingText as NSString).draw(
                at: CGPoint(x: centerX - headingTextSize.width / 2, y: centerY + 15),
                withAttributes: headingAttributes
            )
            
            // Draw direction text below
            let dirAttributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 11),
                .foregroundColor: UIColor.lightGray,
            ]
            (direction as NSString).draw(
                at: CGPoint(x: centerX - 20, y: centerY + 30),
                withAttributes: dirAttributes
            )
        }
        
        return result
    }
}

File: Pekkam-Android/app/src/main/java/com/pekkam/app/ImageOverlayRenderer.kt
package com.pekkam.app

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import kotlin.math.cos
import kotlin.math.sin

object ImageOverlayRenderer {
    fun addWatermark(bitmap: Bitmap): Bitmap {
        val result = bitmap.copy(bitmap.config, true)
        val canvas = Canvas(result)

        val text = "PEKKAM PRØVEVERSJON"
        val paint = Paint().apply {
            textSize = 80f
            color = android.graphics.Color.WHITE
            alpha = 77
            typeface = android.graphics.Typeface.DEFAULT_BOLD
        }

        val textWidth = paint.measureText(text)
        val textHeight = 80f
        val spacing = 300f
        val angle = -45f

        canvas.save()

        var x = -bitmap.width.toFloat()
        while (x < bitmap.width * 2) {
            var y = -bitmap.height.toFloat()
            while (y < bitmap.height * 2) {
                canvas.save()

                val centerX = x + textWidth / 2
                val centerY = y + textHeight / 2

                canvas.translate(centerX, centerY)
                canvas.rotate(angle)
                canvas.translate(-centerX, -centerY)

                canvas.drawText(text, x, y + textHeight, paint)

                canvas.restore()

                y += spacing
            }
            x += spacing
        }

        canvas.restore()
        return result
    }

    fun addCompassOverlay(bitmap: Bitmap, heading: Float, direction: String): Bitmap {
        val result = bitmap.copy(bitmap.config, true)
        val canvas = Canvas(result)

        val compassSize = 300f
        val padding = 40f
        val left = bitmap.width - compassSize - padding
        val top = bitmap.height - compassSize - padding

        // Draw background circle
        val bgPaint = Paint().apply {
            color = android.graphics.Color.BLACK
            alpha = 128
        }
        canvas.drawCircle(left + compassSize / 2, top + compassSize / 2, compassSize / 2, bgPaint)

        // Draw border
        val borderPaint = Paint().apply {
            color = android.graphics.Color.WHITE
            strokeWidth = 4f
            style = Paint.Style.STROKE
        }
        canvas.drawCircle(left + compassSize / 2, top + compassSize / 2, compassSize / 2 - 5, borderPaint)

        val centerX = left + compassSize / 2
        val centerY = top + compassSize / 2
        val radius = (compassSize / 2) - 20f

        // Draw cardinal directions
        val directions = arrayOf("N", "E", "S", "W")
        val angles = arrayOf(0f, 90f, 180f, 270f)

        val dirPaint = Paint().apply {
            color = android.graphics.Color.WHITE
            textSize = 28f
            typeface = android.graphics.Typeface.DEFAULT_BOLD
            textAlign = Paint.Align.CENTER
        }

        for (i in directions.indices) {
            val angle = Math.toRadians(angles[i].toDouble())
            val x = centerX + sin(angle) * radius
            val y = centerY - cos(angle) * radius
            canvas.drawText(directions[i], x.toFloat(), y.toFloat() + 10, dirPaint)
        }

        // Draw needle
        val needleAngle = Math.toRadians(heading.toDouble())
        val needleLength = radius * 0.8f
        val needleEndX = centerX + sin(needleAngle) * needleLength
        val needleEndY = centerY - cos(needleAngle) * needleLength

        val needlePaint = Paint().apply {
            color = android.graphics.Color.RED
            strokeWidth = 6f
        }
        canvas.drawLine(centerX, centerY, needleEndX.toFloat(), needleEndY.toFloat(), needlePaint)

        // Draw heading degree text
        val headingText = String.format("%.0f°", heading)
        val headingPaint = Paint().apply {
            color = android.graphics.Color.WHITE
            textSize = 24f
            typeface = android.graphics.Typeface.DEFAULT_BOLD
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText(headingText, centerX, centerY + 30, headingPaint)

        // Draw direction text
        val dirTextPaint = Paint().apply {
            color = android.graphics.Color.LTGRAY
            textSize = 22f
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText(direction, centerX, centerY + 60, dirTextPaint)

        return result
    }
}
